File: Data Structures & Algorithms/validate-parentheses/submission-2.py
class Solution:
    def isValid(self, s: str) -> bool:
        p_stack = []
        for ch in s:
            if ch == '(' or ch == '{' or ch == '[':
                p_stack.append(ch)
            
            if ch == ')' or ch == '}' or ch == ']':
                if len(p_stack) == 0:
                    return False
                last_paren = p_stack.pop()
                if (
                    ch == ')' and last_paren != '(' or
                    ch == '}' and last_paren != '{' or
                    ch == ']' and last_paren != '['
                ):
                    return False
        
        return len(p_stack) == 0

File: Data Structures & Algorithms/course-schedule/submission-2.cpp
class Course {
public:
    vector<int> adjacencies;
    int prereqs;

    Course() : prereqs(0) {}
};

class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<Course> courses(numCourses);

        for (const vector<int>& prereq : prerequisites) {
            courses[prereq.at(1)].adjacencies.push_back(prereq.at(0));
            courses[prereq.at(0)].prereqs++;
        }

        queue<int> q;
        for (int i = 0; i < numCourses; i++) {
            if (courses[i].prereqs == 0) {
                q.push(i);
            }
        }

        while (!q.empty()) {
            int course_num = q.front();
            q.pop();

            if (courses[course_num].prereqs > 0) {
                continue;
            }

            for (int adj : courses[course_num].adjacencies) {
                q.push(adj);
                courses[adj].prereqs--;
            }
        }

        for (Course& course : courses) {
            cout << course.prereqs << '\n';
            if (course.prereqs != 0) {
                return false;
            }
        }
        
        return true;
    }
};
